crates/polars-stream/src/physical_plan/lower_expr.rs

```diff
@@ -980,6 +980,77 @@ fn lower_exprs_with_ctx(
                 transformed_exprs.push(ctx.expr_arena.add(AExpr::Column(value_key)));
             },
 
+            #[cfg(feature = "diff")]
+            AExpr::Function {
+                input: ref inner_exprs,
+                function: IRFunctionExpr::Diff(null_behavior),
+                options: _,
+            } => {
+                use polars_core::scalar::Scalar;
+                use polars_core::series::ops::NullBehavior;
+
+                assert_eq!(inner_exprs.len(), 2);
+
+                // Transform:
+                //    expr.diff(offset, "ignore")
+                //      ->
+                //    expr - expr.shift(offset)
+
+                let base_expr_ir = &inner_exprs[0];
+                let base_dtype =
+                    base_expr_ir.dtype(&ctx.phys_sm[input.node].output_schema, ctx.expr_arena)?;
+                let offset_expr_ir = &inner_exprs[1];
+                let offset_dtype =
+                    offset_expr_ir.dtype(&ctx.phys_sm[input.node].output_schema, ctx.expr_arena)?;
+
+                let cast_dtype = |dt: &DataType| match *dt {
+                    DataType::UInt8 => Some(DataType::Int16),
+                    DataType::UInt16 => Some(DataType::Int32),
+                    DataType::UInt32 | DataType::UInt64 => Some(DataType::Int64),
+                    _ => None,
+                };
+
+                let mut base = AExprBuilder::new_from_node(base_expr_ir.node());
+                if let Some(cast_dtype) = cast_dtype(base_dtype) {
+                    base = base.cast(cast_dtype, ctx.expr_arena);
+                }
+                let mut offset = AExprBuilder::new_from_node(offset_expr_ir.node());
+                if let Some(cast_dtype) = cast_dtype(offset_dtype) {
```

Comment:
I don't understand why we're passing the offset through `cast_dtype`. I believe the offset should always be simply cast to Int64.